src/ipa_pc/mod.rs

```diff
@@ -736,6 +736,155 @@ impl<G: AffineCurve, D: Digest> PolynomialCommitment<G::ScalarField> for InnerPr
         })
     }
 
+    /// Refers to the "Multi-poly multi-point assertions" section of the "Recursive SNARKs based on Marlin" document
+    fn batch_open_individual_opening_challenges<'a>(
+        ck: &Self::CommitterKey,
+        labeled_polynomials: impl Clone + IntoIterator<Item = &'a LabeledPolynomial<G::ScalarField>>,
+        commitments: impl Clone + IntoIterator<Item = &'a LabeledCommitment<Self::Commitment>>,
+        query_set: &QuerySet<G::ScalarField>,
+        opening_challenges: &dyn Fn(u64) -> G::ScalarField,
+        rands: impl Clone + IntoIterator<Item = &'a Self::Randomness>,
+        rng: Option<&mut dyn RngCore>,
+    ) -> Result<Self::BatchProof, Self::Error>
+    where
+        Self::Randomness: 'a,
+        Self::Commitment: 'a,
+    {
+        let batch_time = start_timer!(|| "Multi poly multi point batching.");
+
+        let mut opening_challenge_counter = 0;
+        
+        // lambda
+        let mut cur_challenge = opening_challenges(opening_challenge_counter);
```

Comment:
Using only a counter (with no random influence) is not secure, unless the function  ``opening_challenges()`` is a stateful random oracle (like a sponge). 

In a generic implementation I would opt to keep with a stateless random oracle. As such a correct implementation of the Fiat-Shamir transform would be as follows:

- We initialize our state by hashing the claim of the (batched) opening proof, state = ``opening_challenges(all commitments, all points, and all claimed values)``. In our setting, we can use this state as the first lambda, ``lambda_1 = state``.
- all other challenges lambda_2, ..., lambda_n are squeezed recursively by ``lambda_i = opening_challenges(lambda_{i-1})``.
- The new opening point x (the "fresh" one) is squeezed by absorbing the new batch poly and the last lambda_n, ``x= opening_challenges(lambda_n, Cbatchedpoly)``.

Alternatively, one can work with counters. But I recommend to do it iteratively, as it comes at (as good as) no extra cost and clearly follows the principle of "absorbing"/"acknowledging" everything that has been "published" before.
Anyway, I again opt for rebranding ``opening_challenges`` to ``random_oracle``.